### DISCUSSION TITLE
Support eager-loading Matrix blocks' owners

### DISCUSSION THREAD

**Discussion by brandonkelly** [craftcms/cms#8445] [Ideas]
Since Craft 3 has a `craft.matrixBlocks()` template function, it would be nice to be able to eager-load Matrix blocks with their respective owners.

Example:

```twig
{% set blocks = craft.matrixBlocks().with(['owner']) %}
```

Related Stack Exchange Q&A: https://craftcms.stackexchange.com/q/21124/9

**Comment by philipzaengle** [comment]
This would be extra nice to have. right now.

**Comment by brandonkelly** [comment]
Little tricky because you could be querying Matrix blocks across a variety of owner element types, and it’s not currently possible to eager-load multiple element types at once from the same eager loading key.

**Comment by hiasl** [comment]
I understand your point, but at least in my projects I would always only load matrix blocks with the owners of the same type (event the same section). Of course there might be other use cases. Maybe it would be possible to enable if the owner type is part of the function call like 

`.withOwner(<type>)`

And unless you believe that it will be possible to solve this generically for all types, maybe this would be a good start.

**Comment by boboldehampsink** [comment]
Came across this and I need this in my life

**Comment by nettum** [comment]
Is this possible now? I'm running a query like
```
MatrixBlock::find()->fieldId(1)->all();
```
and the devtoolbar tells me craft does about 6000 queries.

After updating the query to 
```
MatrixBlock::find()->fieldId(1)->with(['owner'])->all();
```
craft says 150 queries.

But as far as I can see I'm not getting the owner data back?

**Comment by brandonkelly** [comment]
Not yet possible because we need to come up with a good solution for https://github.com/craftcms/cms/issues/1820#issuecomment-479188472 .

**Comment by nettum** [comment]
Ok, thanks for the quick response! :+1:
